python_modules/dagster/dagster/_core/workspace/context.py

```diff
@@ -609,39 +585,50 @@ def has_repository_location_error(self, location_name):
 
     def reload_repository_location(self, name: str) -> None:
         # Can be called from a background thread
+        new = self._load_location(self._location_entry_dict[name].origin)
         with self._lock:
             # Relying on GC to clean up the old location once nothing else
             # is referencing it
-            self._location_entry_dict[name] = self._load_location(
-                self._location_entry_dict[name].origin
-            )
+            self._location_entry_dict[name] = new
 
     def shutdown_repository_location(self, name: str):
         with self._lock:
             self._location_entry_dict[name].origin.shutdown_server()
 
     def reload_workspace(self):
-        # Can be called from a background thread
+        updated_locations = {
+            origin.location_name: self._load_location(origin) for origin in self._origins
+        }
+        self._update_workspace(updated_locations)
+
+    def _update_workspace(self, new_locations: Dict[str, WorkspaceLocationEntry]):
+        # minimize lock time by only holding while swapping data old to new
         with self._lock:
-            self._cleanup_locations()
-            self._load_workspace()
+            previous_events = self._watch_thread_shutdown_events
+            self._watch_thread_shutdown_events = {}
+
+            previous_threads = self._watch_threads
+            self._watch_threads = {}
+
+            previous_locations = self._location_entry_dict
+            self._location_entry_dict = new_locations
 
-    def _cleanup_locations(self):
-        assert self._lock.locked()
-        for _, event in self._watch_thread_shutdown_events.items():
+            # start monitoring for new locations
+            for entry in self._location_entry_dict.values():
```

Comment:
Is there any advantage to keeping the watch thread (if the origin exists as a previous location)?

I could see this doing some sort of intersection based on origin id.